hibernate.go

```diff
@@ -0,0 +1,357 @@
+package cbft
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/couchbase/cbgt"
+	log "github.com/couchbase/clog"
+
+	jp "github.com/buger/jsonparser"
+	"github.com/pkg/errors"
+)
+
+const DefaultIndexActivityStatsSampleIntervalSecs = 30
+const LastAccessedTime = "last_access_time"
+
+var DefaultDoFunc = func(client *http.Client, req *http.Request) (*http.Response, error) {
+	resp, err := client.Do(req)
+	if err != nil {
+		return &http.Response{}, err
+	}
+	return resp, nil
+}
+
+// IndexActivityStats is a struct holding statistics about
+// search indexes.
+type indexActivityStats struct {
+	Url      string        // Ex: "http://10.0.0.1:8095".
+	Start    time.Time     // When we started to get this sample.
+	Duration time.Duration // How long it took to get this sample.
+	Error    error
+	Data     []byte // response from the endpoint.
+}
+
+type doFunc func(*http.Client, *http.Request) (*http.Response, error)
+
+// indexActivityStatsOptions holds options about configuring
+// the polling of search indexes' stats.
+type indexActivityStatsOptions struct {
+	indexActivityStatsSampleInterval time.Duration
+	// The Do field is a function that will execute a HTTP request.
+	// Added for unit testing.
+	Do doFunc
+}
+
+// MonitorIndexActivityStats is used to monitor index activity
+// stats for a specific URL.
+type MonitorIndexActivityStats struct {
+	URL      string // REST URL to monitor
+	SampleCh chan indexActivityStats
+	Options  indexActivityStatsOptions
+	StopCh   chan struct{}
+}
+
+// IndividualIndexActivityStats holds acitivity stats
+// ns each individual index.
+type individualIndexActivityStats struct {
+	LastAccessedTime time.Time `json:"last_accessed_time"`
+}
+
+// separate struct for hibernation policy - last access time, and actions to take
+// buffered channel - push Result into channel?
+// analyse stats from results and apply hibernation policies
+func updateHibernationStatus(mgr *cbgt.Manager,
+	results map[string]individualIndexActivityStats) error {
+	// indexDefs, cas, err := mgr.GetIndexDefs(true) - cannot do this since cas needed later on
+	cfg := mgr.Cfg()
+	indexDefs, cas, err := cbgt.CfgGetIndexDefs(cfg)
```

Comment:
use the manager's instance here?

Reply:
use the same source index definition across updateHibernationStatus as well as unmarshalIndexActivityStats?

Reply:
I did consider using the manager's instance to get index defs but since I'm setting index defs later on, I require the cas value for that, which I will get from CfgSetIndexDefs().

Reply:
In unmarshalIndexActivityStats(), I'm using the manager's instance since I don't require the cas value.

Reply:
there is a standard CAS value defined for overriding the cas.. CFG_CAS_FORCE iirc

Reply:
Thanks, fixed in a recent commit.